internal/dot/graph.go

```diff
@@ -252,6 +318,87 @@ func (dg *Graph) addToGroup(r *Result, id CtorID) {
 	group.Results = append(group.Results, r)
 }
 
+// Prune removes elements from the graph that do not have failed results.
+// Removing elements that do not have failing results makes the graph easier to debug, since non-failing
+// nodes and edges can clutter the graph and don't help the user debug.
+func (dg *Graph) PruneSuccess() {
+	dg.pruneCtors()
+	dg.pruneGroups()
+}
+
+// Remove constructors from the graph that do not have failing Results.
+func (dg *Graph) pruneCtors() {
+	var pruned []*Ctor
+	for _, c := range dg.Ctors {
+		if _, ok := dg.Failed.ctors[c.ID]; ok {
+			pruned = append(pruned, c)
+		} else {
+			dg.pruneCtorParams(c)
+			dg.pruneGroupResults(c)
+			delete(dg.ctorMap, c.ID)
+		}
+	}
+
+	dg.Ctors = pruned
+}
+
+// pruneGroups removes groups from the graph that do not have failing results.
+func (dg *Graph) pruneGroups() {
+	var pruned []*Group
+	for _, g := range dg.Groups {
+		k := g.groupKey()
+		if _, ok := dg.Failed.groups[k]; ok {
+			pruned = append(pruned, g)
+		} else {
+			delete(dg.groupMap, k)
+		}
+	}
+	dg.Groups = pruned
+
+	dg.pruneCtorGroupParams()
+}
+
+// pruneCtorParams removes constructor results that are still referenced in the Params of constructors that consume those results.
+// If the results in the constructor are found in the params of a consuming constructor that result should be removed.
+func (dg *Graph) pruneCtorParams(c *Ctor) {
+	for _, r := range c.Results {
+		if ctors, ok := dg.paramCtorMap[r.nodeKey()]; ok {
+			for _, ctor := range ctors {
+				ctor.removeParams(r)
+			}
+		}
+	}
+}
+
+// pruneCtorGroupParams removes constructor results that are still referenced in the GroupParams of constructors that consume those results.
+func (dg *Graph) pruneCtorGroupParams() {
+	for _, c := range dg.Ctors {
+		var pruned []*Group
+		for _, gp := range c.GroupParams {
+			k := gp.groupKey()
+			if _, ok := dg.groupMap[k]; ok {
+				pruned = append(pruned, gp)
```

Comment:
You're using modifications made to groupMap elsewhere to communicate what
should be deleted here. That works but it's a bit fragile. Information is
being communicated here implicitly through side-effects. For example, if
someone changes the ordering of which prune operation is called when, this
will break and it'll be pretty difficult to debug.

I suggest passing in the list/set of groups to delete/keep in as a parameter.

```
dg.pruneCtorGroupParams(deleted []*Group)
```